File: src/main/java/net/treset/minecraftlauncher/launching/CommandBuilder.java
package net.treset.minecraftlauncher.launching;

import javafx.util.Pair;
import net.hycrafthd.minecraft_authenticator.login.User;
import net.treset.mc_version_loader.launcher.LauncherFeature;
import net.treset.mc_version_loader.launcher.LauncherLaunchArgument;
import net.treset.mc_version_loader.launcher.LauncherManifest;
import net.treset.mc_version_loader.launcher.LauncherVersionDetails;
import net.treset.minecraftlauncher.LauncherApplication;
import net.treset.minecraftlauncher.data.InstanceData;
import net.treset.minecraftlauncher.util.FormatUtil;
import net.treset.minecraftlauncher.util.QuickPlayData;
import net.treset.minecraftlauncher.util.exception.GameCommandException;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class CommandBuilder {
    private static final Logger LOGGER = LogManager.getLogger(CommandBuilder.class);

    private ProcessBuilder processBuilder;
    private InstanceData instanceData;
    private User minecraftUser;
    private QuickPlayData quickPlayData;

    public CommandBuilder(ProcessBuilder processBuilder, InstanceData instanceData, User minecraftUser, QuickPlayData quickPlayData) {
        this.processBuilder = processBuilder;
        this.instanceData = instanceData;
        this.minecraftUser = minecraftUser;
        this.quickPlayData = quickPlayData;
    }

    public void makeStartCommand() throws GameCommandException {
        if(instanceData.getVersionComponents() == null || instanceData.getVersionComponents().isEmpty() || instanceData.getJavaComponent() == null || instanceData.getResourcepacksComponent() == null || instanceData.getLibrariesDir() == null) {
            throw new GameCommandException("Unable to create start command: unmet requirements");
        }

        String assetsIndex = null;
        for(Pair<LauncherManifest, LauncherVersionDetails> v : instanceData.getVersionComponents()) {
            if(v.getValue().getAssets() != null && !v.getValue().getAssets().isBlank()) {
                assetsIndex = v.getValue().getAssets();
                break;
            }
        }
        if(assetsIndex == null) {
            throw new GameCommandException("Unable to create start command: unable to determine asset index");
        }

        String mainClass = null;
        for(Pair<LauncherManifest, LauncherVersionDetails> v : instanceData.getVersionComponents()) {
            if(v.getValue().getMainFile() != null && !v.getValue().getMainFile().isBlank()) {
                mainClass = v.getValue().getMainClass();
                break;
            }
        }
        if(mainClass == null) {
            throw new GameCommandException("Unable to create start command: unable to determine main class");
        }

        List<String> libraries = new ArrayList<>();
        for(Pair<LauncherManifest, LauncherVersionDetails> v : instanceData.getVersionComponents()) {
            for(String l : v.getValue().getLibraries()) {
                libraries.add(instanceData.getLibrariesDir() + l);
            }
        }
        if(libraries.isEmpty()) {
            throw new GameCommandException("Unable to create start command: unable to determine libraries");
        }
        for(Pair<LauncherManifest, LauncherVersionDetails> v : instanceData.getVersionComponents()) {
            if(v.getValue().getMainFile() == null ||v.getValue().getMainFile().isBlank()) {
                throw new GameCommandException("Unable to create start command: unable to determine main file: version=" + v.getKey().getId());
            }
            libraries.add(v.getKey().getDirectory() + v.getValue().getMainFile());
        }

        String resX = null;
        String resY = null;
        for(LauncherFeature f : instanceData.getInstance().getValue().getFeatures()) {
            if(f.getFeature().equals("resolution_x")) {
                resX = f.getValue();
            }
            if(f.getFeature().equals("resolution_y")) {
                resY = f.getValue();
            }
        }

        File gameDir = new File(instanceData.getGameDataDir());
        if(!gameDir.isDirectory()) {
            throw new GameCommandException("Unable to create start command: game directory is not a directory: directory=" + gameDir.getAbsolutePath());
        }

        processBuilder.directory(gameDir);
        processBuilder.command(new ArrayList<>());
        processBuilder.command().add(FormatUtil.absoluteFilePath(instanceData.getJavaComponent().getDirectory(), "bin", "java"));
        try {
            appendArguments(processBuilder, instanceData.getInstance().getValue().getJvm_arguments(), instanceData, minecraftUser, instanceData.getGameDataDir(), instanceData.getAssetsDir(), assetsIndex, libraries, resX, resY, quickPlayData);
        } catch(GameCommandException e) {
            throw new GameCommandException("Unable to create start command: unable to append instance jvm arguments: version=" + instanceData.getInstance().getKey().getId(), e);
        }
        for(Pair<LauncherManifest, LauncherVersionDetails> v : instanceData.getVersionComponents()) {
            try {
                appendArguments(processBuilder, v.getValue().getJvmArguments(), instanceData, minecraftUser, instanceData.getGameDataDir(), instanceData.getAssetsDir(), assetsIndex, libraries, resX, resY, quickPlayData);
            } catch(GameCommandException e) {
                throw new GameCommandException("Unable to create start command: unable to append jvm arguments: version=" + v.getKey().getId(), e);
            }
        }
        processBuilder.command().add(mainClass);
        for(Pair<LauncherManifest, LauncherVersionDetails> v : instanceData.getVersionComponents()) {
            try {
                appendArguments(processBuilder, v.getValue().getGameArguments(), instanceData, minecraftUser, instanceData.getGameDataDir(), instanceData.getAssetsDir(), assetsIndex, libraries, resX, resY, quickPlayData);
            } catch(GameCommandException e) {
                throw new GameCommandException("Unable to create start command: unable to append game arguments: version=" + v.getKey().getId(), e);
            }
        }

        LOGGER.info("Created start command, instance=" + instanceData.getInstance().getKey().getId());
    }

    private void appendArguments(ProcessBuilder pb, List<LauncherLaunchArgument> args, InstanceData instanceData, User minecraftUser, String gameDataDir, String assetsDir, String assetsIndex, List<String> libraries, String resX, String resY, QuickPlayData quickPlayData) throws GameCommandException {
        List<GameCommandException> exceptionQueue = new ArrayList<>();
        for(LauncherLaunchArgument a : args) {
            try {
                appendArgument(pb, instanceData, minecraftUser, gameDataDir, assetsDir, assetsIndex, libraries, resX, resY, a, quickPlayData);
            } catch (GameCommandException e) {
                exceptionQueue.add(e);
                LOGGER.warn("Unable to append arguments: unable to append argument: argument=" + a);
            }
        }
        if(!exceptionQueue.isEmpty()) {
            throw new GameCommandException("Unable to append " + exceptionQueue.size() + " arguments", exceptionQueue.get(0));
        }
    }

    private void appendArgument(ProcessBuilder pb, InstanceData instanceData, User minecraftUser, String gameDataDir, String assetsDir, String assetsIndex, List<String> libraries, String resX, String resY, LauncherLaunchArgument a, QuickPlayData quickPlayData) throws GameCommandException {
        if(a.isActive(instanceData.getInstance().getValue().getFeatures())) {
            Map<String, String> replacements = new HashMap<>();
            List<GameCommandException> exceptionQueue = new ArrayList<>();
            for (String r : a.getReplacementValues()) {
                try {
                    String replacement = getReplacement(r, gameDataDir, instanceData.getJavaComponent().getDirectory(), assetsDir, instanceData.getResourcepacksComponent().getDirectory(), assetsIndex, libraries, minecraftUser, LauncherApplication.stringLocalizer.getFormatted("game.version_name", LauncherApplication.stringLocalizer.get("launcher.slug"), LauncherApplication.stringLocalizer.get("launcher.version"), instanceData.getInstance().getKey().getId().substring(0, 3), instanceData.getInstance().getKey().getId().substring(instanceData.getInstance().getKey().getId().length()-2)), LauncherApplication.stringLocalizer.getFormatted("game.version_type", instanceData.getInstance().getKey().getName()), resX, resY, quickPlayData);
                    replacements.put(r, replacement);
                } catch (GameCommandException e) {
                    exceptionQueue.add(e);
                    LOGGER.warn("Unable to append argument: unable to replace variable: argument=" + a.getArgument() + ", variable=" + r, e);
                }
            }
            if(!exceptionQueue.isEmpty()) {
                throw new GameCommandException("Unable to append argument: unable to replace " + exceptionQueue.size() + " variables: argument=" + a.getArgument(), exceptionQueue.get(0));
            }
            a.replace(replacements);
            if(!a.isFinished()) {
                throw new GameCommandException("Unable to append argument: unable to replace all variables: argument=" + a);
            }
            pb.command().add(a.getParsedArgument());
        }
    }

    private String getReplacement(String key, String gameDir, String javaDir, String assetsDir, String resourcepackDir, String assetsIndex, List<String> libraries, User minecraftUser, String versionName, String versionType, String resX, String resY, QuickPlayData quickPlayData) throws GameCommandException {
        switch(key) {
            case "natives_directory" -> {
                return javaDir + "lib";
            }
            case "launcher_name" -> {
                return LauncherApplication.stringLocalizer.get("launcher.name");
            }
            case "launcher_version" -> {
                return LauncherApplication.stringLocalizer.get("launcher.version");
            }
            case "classpath" -> {
                StringBuilder sb = new StringBuilder();
                for(String l : libraries) {
                    sb.append(l).append(";");
                }
                return sb.substring(0, sb.length() - 1);
            }
            case "auth_player_name" -> {
                return minecraftUser.name();
            }
            case "version_name" -> {
                return versionName;
            }
            case "game_directory" -> {
                return gameDir;
            }
            case "resourcepack_directory" -> {
                return resourcepackDir;
            }
            case "assets_root" -> {
                return assetsDir;
            }
            case "assets_index_name" -> {
                return assetsIndex;
            }
            case "auth_uuid" -> {
                return minecraftUser.uuid();
            }
            case "auth_xuid" -> {
                return minecraftUser.xuid();
            }
            case "auth_access_token" -> {
                return minecraftUser.accessToken();
            }
            case "clientid" -> {
                return minecraftUser.clientId();
            }
            case "user_type" -> {
                return minecraftUser.type();
            }
            case "version_type" -> {
                return versionType;
            }
            case "resolution_width" -> {
                return resX;
            }
            case "resolution_height" -> {
                return resY;
            }
            case "quickPlayPath" -> {
                return "quickPlay/log.json";
            }
            case "quickPlaySingleplayer" -> {
                return quickPlayData != null && quickPlayData.getType() == QuickPlayData.Type.WORLD ? quickPlayData.getName() : "";
            }
            case "quickPlayMultiplayer" -> {
                return quickPlayData != null && quickPlayData.getType() == QuickPlayData.Type.SERVER ? quickPlayData.getName() : "";
            }
            case "quickPlayRealms" -> {
                return quickPlayData != null && quickPlayData.getType() == QuickPlayData.Type.REALM ? quickPlayData.getName() : "";
            }

            default -> throw new GameCommandException("Unknown environment variable: key=" + key);
        }
    }

    public ProcessBuilder getProcessBuilder() {
        return processBuilder;
    }

    public void setProcessBuilder(ProcessBuilder processBuilder) {
        this.processBuilder = processBuilder;
    }

    public InstanceData getInstanceData() {
        return instanceData;
    }

    public void setInstanceData(InstanceData instanceData) {
        this.instanceData = instanceData;
    }

    public User getMinecraftUser() {
        return minecraftUser;
    }

    public void setMinecraftUser(User minecraftUser) {
        this.minecraftUser = minecraftUser;
    }

    public QuickPlayData getQuickPlayData() {
        return quickPlayData;
    }

    public void setQuickPlayData(QuickPlayData quickPlayData) {
        this.quickPlayData = quickPlayData;
    }
}

File: client/composeApp/src/commonMain/kotlin/net/treset/treelauncher/backend/launching/CommandBuilder.kt
package net.treset.treelauncher.backend.launching

import io.github.oshai.kotlinlogging.KotlinLogging
import net.hycrafthd.minecraft_authenticator.login.User
import net.treset.treelauncher.backend.data.InstanceData
import net.treset.treelauncher.backend.data.LauncherLaunchArgument
import net.treset.treelauncher.backend.util.QuickPlayData
import net.treset.treelauncher.backend.util.exception.GameCommandException
import net.treset.treelauncher.backend.util.file.LauncherFile
import net.treset.treelauncher.localization.strings
import java.io.File

class CommandContext(
    val offline: Boolean,
    val minecraftUser: User?,
    val gameDataDir: String,
    val resourcepacksDir: String,
    val assetsDir: String,
    val assetsIndex: String,
    val libraries: List<String>,
    val librariesDir: String,
    val nativesDir: String,
    val versionName: String,
    val versionType: String,
    val resX: String?,
    val resY: String?,
    val quickPlayData: QuickPlayData?
)

class CommandBuilder(
    var processBuilder: ProcessBuilder,
    private var instanceData: InstanceData,
    var offline: Boolean,
    var minecraftUser: User?,
    private var quickPlayData: QuickPlayData?
) {
    @Throws(GameCommandException::class)
    fun makeStartCommand() {
        if (instanceData.versionComponents.isEmpty()) {
            throw GameCommandException("Unable to create start command: no version available")
        }
        var assetsIndex: String? = null
        for (v in instanceData.versionComponents) {
            if (v.second.assets?.isNotBlank() == true) {
                assetsIndex = v.second.assets
                break
            }
        }
        if (assetsIndex == null) {
            throw GameCommandException("Unable to create start command: unable to determine asset index")
        }

        var mainClass: String? = null
        for (v in instanceData.versionComponents) {
            if (v.second.mainClass.isNotBlank()) {
                mainClass = v.second.mainClass
                break
            }
        }
        if (mainClass == null) {
            throw GameCommandException("Unable to create start command: unable to determine main class")
        }

        val libraries: MutableList<String> = mutableListOf()
        for (v in instanceData.versionComponents) {
            for (l in v.second.libraries) {
                val library = LauncherFile.of(instanceData.librariesDir, l).absolutePath
                if(libraries.any { it == library }) continue
                libraries.add(library)
            }
        }
        if (libraries.isEmpty()) {
            throw GameCommandException("Unable to create start command: unable to determine libraries")
        }

        for (v in instanceData.versionComponents) {
            if (v.second.mainFile?.isNotBlank() == true) {
                libraries.add(LauncherFile.of(v.first.directory, v.second.mainFile!!).absolutePath)
            }
        }

        val natives: MutableList<String> = mutableListOf()
        for(v in instanceData.versionComponents) {
            if(v.second.natives?.isNotBlank() == true) {
                natives.add(LauncherFile.of(v.first.directory, v.second.natives!!).absolutePath)
            }
        }
        if(natives.isEmpty()) {
            natives.add(LauncherFile.of(instanceData.javaComponent.directory, "libs").absolutePath)
        }
        val nativesDir = natives.joinToString(File.pathSeparator)

        var resX: String? = null
        var resY: String? = null
        for (f in instanceData.instance.second.features) {
            if (f.feature == "resolution_x") {
                resX = f.value
            }
            if (f.feature == "resolution_y") {
                resY = f.value
            }
        }
        val gameDir: LauncherFile = instanceData.gameDataDir
        if (!gameDir.isDirectory()) {
            throw GameCommandException("Unable to create start command: game directory is not a directory: directory=${gameDir.absolutePath}")
        }

        processBuilder.directory(gameDir)
        processBuilder.command(mutableListOf())
        processBuilder.command().add(LauncherFile.of(instanceData.javaComponent.directory, "bin", "java").path)
        val argOrder: MutableList<Array<LauncherLaunchArgument>> = mutableListOf(instanceData.instance.second.jvmArguments.toTypedArray())
        for (v in instanceData.versionComponents) {
            argOrder.add(v.second.jvmArguments.toTypedArray())
        }
        argOrder.add(arrayOf(
            LauncherLaunchArgument(mainClass)
        ))
        for (v in instanceData.versionComponents) {
            argOrder.add(v.second.gameArguments.toTypedArray())
        }

        val context = CommandContext(
            offline,
            minecraftUser,
            instanceData.gameDataDir.absolutePath,
            instanceData.resourcepacksComponent.directory,
            instanceData.assetsDir.absolutePath,
            assetsIndex,
            libraries,
            instanceData.librariesDir.absolutePath,
            nativesDir,
            strings().game.versionName(instanceData),
            strings().game.versionType(instanceData),
            resX,
            resY,
            quickPlayData
        )

        try {
            appendArguments(
                processBuilder,
                argOrder.toTypedArray(),
                context
            )
        } catch (e: GameCommandException) {
            throw GameCommandException("Unable to create start command: unable to append arguments", e)
        }
        LOGGER.info { "Created start command, instance=${instanceData.instance.first.id}" }
    }

    @Throws(GameCommandException::class)
    private fun appendArguments(
        pb: ProcessBuilder,
        argOrder: Array<Array<LauncherLaunchArgument>>,
        context: CommandContext
    ) {
        for(args in argOrder) {
            appendArguments(
                pb,
                args,
                context,
            )
        }
    }

    @Throws(GameCommandException::class)
    private fun appendArguments(
        pb: ProcessBuilder,
        args: Array<LauncherLaunchArgument>,
        context: CommandContext
    ) {
        val exceptionQueue: MutableList<GameCommandException> = mutableListOf()
        for (a in args) {
            try {
                appendArgument(
                    pb,
                    a,
                    context
                )
            } catch (e: GameCommandException) {
                exceptionQueue.add(e)
                LOGGER.warn { "Unable to append arguments: unable to append argument: argument=$a" }
            }
        }
        if (exceptionQueue.isNotEmpty()) {
            throw GameCommandException("Unable to append ${exceptionQueue.size} arguments", exceptionQueue[0])
        }
    }

    @Throws(GameCommandException::class)
    private fun appendArgument(
        pb: ProcessBuilder,
        a: LauncherLaunchArgument,
        context: CommandContext
    ) {
        if (a.isActive(instanceData.instance.second.features)) {
            val replacements: MutableMap<String, String> = mutableMapOf()
            val exceptionQueue: MutableList<GameCommandException> = mutableListOf()
            for (r in a.replacementValues) {
                try {
                    val replacement = getReplacement(
                        r,
                        context
                    )
                    replacements[r] = replacement
                } catch (e: GameCommandException) {
                    exceptionQueue.add(e)
                    LOGGER.warn(e) { "Unable to append argument: unable to replace variable: argument=${a.argument}, variable=${r}" }
                    replacements[r] = ""
                }
            }
            if (exceptionQueue.isNotEmpty()) {
                LOGGER.error(exceptionQueue[0]) { "Unable to append argument: unable to replace ${exceptionQueue.size} variables: argument=${a.argument}" }
            }
            a.replace(replacements)
            if (!a.isFinished) {
                throw GameCommandException("Unable to append argument: unable to replace all variables: argument=${a.argument}")
            }
            pb.command().add(a.parsedArgument)
        }
    }

    @Throws(GameCommandException::class)
    private fun getReplacement(
        key: String,
        context: CommandContext
    ): String {
        return when (key) {
            "natives_directory" -> {
                context.nativesDir
            }

            "launcher_name" -> {
                strings().launcher.name()
            }

            "launcher_version" -> {
                strings().launcher.version()
            }

            "classpath" -> {
                val sb = StringBuilder()
                for (l in context.libraries) {
                    sb.append(l).append(";")
                }
                sb.substring(0, sb.length - 1)
            }

            "classpath_separator" -> {
                File.pathSeparator
            }

            "library_directory" -> {
                context.librariesDir
            }

            "auth_player_name" -> {
                if(context.minecraftUser != null && !context.offline) {
                    context.minecraftUser.name()
                } else {
                    "LocalPlayer"
                }
            }

            "version_name" -> {
                context.versionName
            }

            "game_directory" -> {
                context.gameDataDir
            }

            "resourcepack_directory" -> {
                context.resourcepacksDir
            }

            "assets_root" -> {
                context.assetsDir
            }

            "game_assets" -> {
                context.assetsDir
            }

            "assets_index_name" -> {
                context.assetsIndex
            }

            "auth_uuid" -> {
                if(context.minecraftUser != null && !context.offline) {
                    context.minecraftUser.uuid()
                } else {
                    "00000000000000000000000000000000"
                }
            }

            "auth_xuid" -> {
                if(context.minecraftUser != null && !context.offline) {
                    context.minecraftUser.xuid()
                } else {
                    ""
                }
            }

            "auth_access_token" -> {
                if(context.minecraftUser != null && !context.offline) {
                    context.minecraftUser.accessToken()
                } else {
                    ""
                }
            }

            "clientid" -> {
                if(context.minecraftUser != null && !context.offline) {
                    context.minecraftUser.clientId()
                } else {
                    ""
                }
            }

            "user_type" -> {
                if(context.minecraftUser != null && !context.offline) {
                    context.minecraftUser.type()
                } else {
                    ""
                }
            }

            "auth_session" -> {
                if(context.minecraftUser != null && !context.offline) {
                    "token:${context.minecraftUser.accessToken()}:${context.minecraftUser.uuid()}"
                } else {
                    "token::00000000000000000000000000000000"
                }
            }

            "version_type" -> {
                context.versionType
            }

            "resolution_width" -> {
                context.resX ?: ""
            }

            "resolution_height" -> {
                context.resY ?: ""
            }

            "quickPlayPath" -> {
                "quickPlay/log.json"
            }

            "quickPlaySingleplayer" -> {
                if (context.quickPlayData?.type == QuickPlayData.Type.WORLD) context.quickPlayData.name else ""
            }

            "quickPlayMultiplayer" -> {
                if (context.quickPlayData?.type == QuickPlayData.Type.SERVER) context.quickPlayData.name else ""
            }

            "quickPlayRealms" -> {
                if (context.quickPlayData?.type === QuickPlayData.Type.REALM) context.quickPlayData.name else ""
            }

            else -> throw GameCommandException("Unknown environment variable: key=$key")
        }
    }

    companion object {
        private val LOGGER = KotlinLogging.logger {}
    }
}
